### DISCUSSION TITLE
How to automatically use volumes from an old node on a new one

### DISCUSSION THREAD

**Discussion by sarg3nt** [longhorn/longhorn#8010] [Q&A]
In short.  We are using an immutable operating system in which we do not upgrade nodes but replace them.  I have longhorn data being stored in a separate disk that is remounted once the new node comes up.  The new node has the same node name and FQDN but a different IP and MachineID
Longhorn is not using the data on the volume that already exists but is marking the volume as failed and rebuilding it.

- I'm not seeing a setting that would fix this for us.  
- Or maybe I changed a setting that is causing this issue?  
- Or is this not supported in our config?

This is a problem for our upgrade strategy as large volumes take a long time to rebuild so it will REALLY slow cluster upgrades.

See screenshot below

![image](https://github.com/longhorn/longhorn/assets/3372486/7ad801cd-3708-491f-8c65-ad2e8ce8cc26)

Here's the logs from the volume rebuild

| Name | Kind | Source | Type | First Time | Last Time | Reason | Message |
|-- | -- | -- | -- | -- | -- | -- | -- |
| pvc-6cb86bb3-f0ce-4280-998f-3b852ef24d73-e-0 | Engine | longhorn-engine-controller | Normal | 20 minutes ago | 20 minutes ago | Rebuilt | Replica  pvc-6cb86bb3-f0ce-4280-998f-3b852ef24d73-r-13a99043 with Address  192.168.4.6:10040 has been rebuilt for volume pvc-6cb86bb3-f0ce-4280-998f-3b852ef24d73 |
| pvc-6cb86bb3-f0ce-4280-998f-3b852ef24d73-e-0 | Engine | longhorn-engine-controller | Normal | 33 minutes ago | 33 minutes ago | Rebuilding | Detected rebuilding replica pvc-6cb86bb3-f0ce-4280-998f-3b852ef24d73-r-13a99043 (192.168.4.6:10040) |
| pvc-6cb86bb3-f0ce-4280-998f-3b852ef24d73-e-0 | Engine | longhorn-engine-controller | Normal | 33 minutes ago | 33 minutes ago | Rebuilding | Start  rebuilding replica pvc-6cb86bb3-f0ce-4280-998f-3b852ef24d73-r-13a99043  with Address 192.168.4.6:10040 for normal engine  pvc-6cb86bb3-f0ce-4280-998f-3b852ef24d73-e-0 and volume  pvc-6cb86bb3-f0ce-4280-998f-3b852ef24d73 |
| pvc-6cb86bb3-f0ce-4280-998f-3b852ef24d73-r-13a99043 | Replica | longhorn-replica-controller | Normal | 33 minutes ago | 33 minutes ago | Start | Starts pvc-6cb86bb3-f0ce-4280-998f-3b852ef24d73-r-13a99043 |
| pvc-6cb86bb3-f0ce-4280-998f-3b852ef24d73-e-0 | Engine | longhorn-engine-controller | Normal | 41 minutes ago | 41 minutes ago | Delete | Removing unknown replica tcp://192.168.4.17:10041 in mode RW from engine |
| pvc-6cb86bb3-f0ce-4280-998f-3b852ef24d73-e-0 | Engine | longhorn-engine-controller | Warning | 41 minutes ago | 41 minutes ago | FailedDeleting | Failed  to remove unknown replica tcp://192.168.4.17:10041 in mode RW from  engine: proxyServer=192.168.5.10:8501 destination=192.168.5.10:10051:  failed to remove replica tcp://192.168.4.17:10041 for volume: rpc error:  code = Internal desc = grpc: error while marshaling: proto: Marshal  called with nil |
| pvc-6cb86bb3-f0ce-4280-998f-3b852ef24d73 | Volume | longhorn-volume-controller | Normal | 41 minutes ago | 41 minutes ago | Degraded | volume pvc-6cb86bb3-f0ce-4280-998f-3b852ef24d73 became degraded |

> **Note**:  I tried getting into the Longhorn Slack channel but when I got the email it kept saying "It looks like there isn’t an account on Cloud Native Computing Foundation tied to this email address."  I have a Linux foundation account tied to that email, not sure what else to do there.

**Comment by derekbit** [comment]
Currently, volume rebuilding is evitable when you are replacing a node where a replica is ruining on.
Is it possible to detach your volume before the replacement?

**Reply by sarg3nt** [reply]
What do you mean by "detach the volume"?  Are we talking the disk attached to the node or the volumes in k8s, and if so, how would I go about doing that, doesn't longhorn take care of that?  Or do you mean spin down the workloads?

**Reply by sarg3nt** [reply]
I haven't timed it yet, but is there any speed increase for keeping the longhorn volume around, i.e. persistent vs a fresh volume each time in terms of the rebuild speed / bandwidth?

**Reply by sarg3nt** [reply]
Oh, I see, even just restarting a node causes a rebuild.
So yeah, same question does keeping the Longhorn volume on the node around get us anywhere in terms of speed or should I just create a new one each time?

**Reply by derekbit** [reply]
> What do you mean by "detach the volume"? Are we talking the disk attached to the node or the volumes in k8s, and if so, how would I go about doing that, doesn't longhorn take care of that? Or do you mean spin down the workloads?

Yes, I mean spin down the workloads. After scaling down the volume, the volume will be detached.

> I haven't timed it yet, but is there any speed increase for keeping the longhorn volume around, i.e. persistent vs a fresh volume each time in terms of the rebuild speed / bandwidth?

We have a [fast replica rebuilding](https://longhorn.io/docs/1.6.0/advanced-resources/rebuilding/fast-replica-rebuild/). You can check if it is helpful in your environment. However, if the volume head contains huge data, the rebuilding still takes time. 


> fresh volume each time

Sorry. What does it mean? Don't you need to persistent your data?

**Reply by sarg3nt** [reply]
We have 3 nodes, each with a persistent volume that houses the longhorn data, i.e. `/var/lib/rancher/longhorn` is a volume such as `/dev/sdc`. That volume can either be persisted or ephemeral as it is a VSphere volume.  If it is ephemeral than Longhorn rebuilds the data when it comes up, if it is persistent, containing the longhorn data from the previous node, it also seems to rebuild the data.  So my question is, does it matter?  Should we just start with a blank  `/var/lib/rancher/longhorn`  volume for each node or should we persist the data?
And yeah, I have fast replica rebuilding turned on.

**Comment by R0flcopt3r** [comment]
We have the same sort of problem. Even a reboot triggers a reinstall in our case. Power outages etc will sometimes cause all nodes to go down, meaning we lose all data... We have backup strategies in place, so it's not really a problem. It would however be very nice if longhorn could reuse old data.
